modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Run.scala

```diff
@@ -0,0 +1,152 @@
+                file = parsed.configFile
+                goodSink = initAttributedSink(blocker, file.auth, file.good, mkGoodSink)
+                piiSink = file.pii.map(out => initAttributedSink(blocker, file.auth, out, mkPiiSink))
+                badSink = file.bad match {
+                  case Output.FileSystem(path) =>
+                    Sink.fileSink[F](path, blocker)
+                  case _ =>
+                    mkBadSink(blocker, file.auth, file.bad)
+                }
+                exit <- 
+                  (file.auth, file.input) match {
+                    case (_, p: Input.FileSystem) => 
+                      val env = Environment
+                        .make[F, Array[Byte]](
+                          blocker,
+                          ec,
+                          parsed,
+                          Source.filesystem[F](blocker, p.dir),
+                          goodSink,
+                          piiSink,
+                          badSink,
+                          _.void,
+                          identity,
+                          processor
+                        )
+                      runEnvironment[F, Array[Byte]](env, false)
+                    case _ =>
+                      val env = Environment
+                        .make[F, A](
+                          blocker,
+                          ec,
+                          parsed,
+                          mkSource(blocker, file.auth, file.input),
+                          goodSink,
+                          piiSink,
+                          badSink,
+                          checkpointer,
+                          getPayload,
+                          processor
+                        )
+                      runEnvironment[F, A](env, ordered)
+                  }
+              } yield exit
+            }
+        ).flatten
+      case Left(error) =>
+        Sync[F].delay(System.err.println(error)) >> Sync[F].pure(ExitCode.Error)
+    }
+
+  private def initAttributedSink[F[_]: Concurrent: ContextShift: Timer](
+    blocker: Blocker,
+    auth: Authentication,
+    output: Output,
+    mkGoodSink: (Blocker, Authentication, Output) => Resource[F, AttributedByteSink[F]],
+  ): Resource[F, AttributedByteSink[F]] =
+    output match {
+      case Output.FileSystem(path) =>
+        Sink.fileSink[F](path, blocker).map(sink => row => sink(row.data))
+      case _ =>
+        mkGoodSink(blocker, auth, output)
+    }
+
+  private def runEnvironment[F[_]: ConcurrentEffect: ContextShift: Parallel: Timer, A](
+    environment: Resource[F, Environment[F, A]],
+    ordered: Boolean
+  ): F[ExitCode] =
+    environment.use { env =>
+      val log = Logger[F].info("Running enrichment stream")
+      val enrich = Enrich.run[F, A](env, ordered)
+      val updates = Assets.run[F, A](env)
+      val reporting = env.metrics.report
+      val flow = enrich.merge(updates).merge(reporting)
+      log >> flow.compile.drain.attempt.flatMap {
```

Comment:
I've been using `attempt.flatMap` pattern extensively until recently when I realised there are plenty of `recover`-like function that do exactly the same.